File: typescript/agent.ts
/**
 * Compliance checker agent — TypeScript example.
 *
 * Validates deployment compliance. After this agent completes,
 * the workflow pauses for human CAB approval.
 *
 * Usage:
 *   export AXME_API_KEY="<agent-key>"
 *   npx tsx agent.ts
 */

import { AxmeClient } from "@axme/axme";

const AGENT_ADDRESS = "compliance-checker-demo";

async function handleIntent(client: AxmeClient, intentId: string) {
  const intentData = await client.getIntent(intentId);
  const intent = intentData.intent ?? intentData;
  let payload = intent.payload ?? {};
  if (payload.parent_payload) {
    payload = payload.parent_payload;
  }

  const service = payload.service ?? "unknown";
  const version = payload.version ?? "unknown";
  const env = payload.environment ?? "unknown";
  const risk = payload.risk_level ?? "unknown";

  console.log(`  Checking compliance for ${service} v${version} -> ${env}...`);
  await new Promise((r) => setTimeout(r, 1000));

  console.log(`  Risk level: ${risk}`);
  console.log(`  Validating security policies...`);
  await new Promise((r) => setTimeout(r, 1000));

  console.log(`  Checking rollback plan...`);
  await new Promise((r) => setTimeout(r, 1000));

  const passed = risk !== "critical";
  const result = {
    action: "complete",
    passed,
    checks: {
      security_policy: "passed",
      rollback_plan: "passed",
      dependency_scan: "passed",
      risk_assessment: passed ? "passed" : "failed",
    },
    recommendation: passed ? "approve" : "reject",
  };

  await client.resumeIntent(intentId, result, { ownerAgent: "compliance-checker-demo" });
  const checkStatus = passed ? "PASSED" : "FAILED";
  console.log(`  Compliance check ${checkStatus}. Recommendation: ${result.recommendation}`);
  console.log(`  Workflow now waits for human CAB approval.`);
  console.log(`  To approve: axme tasks approve <intent_id>`);
}

async function main() {
  const apiKey = process.env.AXME_API_KEY;
  if (!apiKey) {
    console.error("Error: AXME_API_KEY not set.");
    process.exit(1);
  }

  const client = new AxmeClient({ apiKey });

  console.log(`Agent listening on ${AGENT_ADDRESS}...`);
  console.log("Waiting for intents (Ctrl+C to stop)\n");

  for await (const delivery of client.listen(AGENT_ADDRESS)) {
    const intentId = delivery.intent_id;
    const status = delivery.status;
    if (!intentId) continue;
    if (["DELIVERED", "CREATED", "IN_PROGRESS"].includes(status)) {
      console.log(`[${status}] Intent received: ${intentId}`);
      try {
        await handleIntent(client, intentId);
      } catch (e) {
        console.error(`  Error: ${e}`);
      }
    }
  }
}

main().catch(console.error);

File: java/Agent.java
/*
 * Compliance checker agent — Java example.
 *
 * Fetches an intent by ID, runs compliance checks, and resumes with result.
 * After this agent completes, the workflow pauses for human CAB approval.
 *
 * Usage:
 *   export AXME_API_KEY="<agent-key>"
 *   javac -cp axme-sdk.jar Agent.java
 *   java -cp .:axme-sdk.jar Agent <intent_id>
 */

import dev.axme.sdk.AxmeClient;
import dev.axme.sdk.AxmeClientConfig;
import dev.axme.sdk.RequestOptions;
import java.util.Map;

public class Agent {
    public static void main(String[] args) throws Exception {
        if (args.length < 1) {
            System.err.println("Usage: java Agent <intent_id>");
            System.exit(1);
        }

        String apiKey = System.getenv("AXME_API_KEY");
        if (apiKey == null || apiKey.isEmpty()) {
            System.err.println("Error: AXME_API_KEY not set.");
            System.exit(1);
        }

        String intentId = args[0];
        var client = new AxmeClient(AxmeClientConfig.forCloud(apiKey));

        System.out.println("Processing intent: " + intentId);

        var intentData = client.getIntent(intentId, new RequestOptions());
        @SuppressWarnings("unchecked")
        var intent = (Map<String, Object>) intentData.getOrDefault("intent", intentData);
        @SuppressWarnings("unchecked")
        var payload = (Map<String, Object>) intent.getOrDefault("payload", Map.of());
        if (payload.containsKey("parent_payload")) {
            @SuppressWarnings("unchecked")
            var pp = (Map<String, Object>) payload.get("parent_payload");
            payload = pp;
        }

        String service = (String) payload.getOrDefault("service", "unknown");
        String version = (String) payload.getOrDefault("version", "unknown");
        String env = (String) payload.getOrDefault("environment", "unknown");
        String risk = (String) payload.getOrDefault("risk_level", "unknown");

        System.out.println("  Checking compliance for " + service + " v" + version + " -> " + env + "...");
        Thread.sleep(1000);

        System.out.println("  Risk level: " + risk);
        System.out.println("  Validating security policies...");
        Thread.sleep(1000);

        System.out.println("  Checking rollback plan...");
        Thread.sleep(1000);

        boolean passed = !risk.equals("critical");
        String riskAssessment = passed ? "passed" : "failed";
        String recommendation = passed ? "approve" : "reject";

        var result = Map.<String, Object>of(
            "action", "complete",
            "passed", passed,
            "checks", Map.of(
                "security_policy", "passed",
                "rollback_plan", "passed",
                "dependency_scan", "passed",
                "risk_assessment", riskAssessment
            ),
            "recommendation", recommendation
        );

        client.resumeIntent(intentId, result, new RequestOptions());
        String status = passed ? "PASSED" : "FAILED";
        System.out.println("  Compliance check " + status + ". Recommendation: " + recommendation);
        System.out.println("  Workflow now waits for human CAB approval.");
        System.out.println("  To approve: axme tasks approve <intent_id>");
    }
}
